source/tests/unit/ni_fake_non_ivi_service_tests.cpp

```diff
@@ -1622,6 +1622,72 @@ TEST_F(NiFakeNonIviServiceTests, SessionAlreadyInCrossDriverSessionRepository_Ge
 
   EXPECT_EQ(original_session_id, cross_driver_resource_repository_->resolve_session_id(CROSS_DRIVER_HANDLE));
 }
+
+TEST_F(NiFakeNonIviServiceTests, InitWithReturnedSession_AccessSession_ReturnsInitializeSessionHandle)
+{
+  constexpr auto SESSION_NAME = "session";
+  constexpr auto SESSION_HANDLE = 0x1234UL;
+  EXPECT_CALL(library_, InitWithReturnedSession(StrEq(SESSION_NAME)))
+      .WillOnce(Return(SESSION_HANDLE));
+  ::grpc::ServerContext context;
+  InitWithReturnedSessionRequest request;
+  InitWithReturnedSessionResponse response;
+  request.set_handle_name(SESSION_NAME);
+  service_.InitWithReturnedSession(&context, &request, &response);
+
+  const auto accessed_handle = resource_repository_->access_session(response.handle().id(), "");
+
+  EXPECT_EQ(kDriverSuccess, response.status());
+  EXPECT_EQ(SESSION_HANDLE, accessed_handle);
+}
+
+TEST_F(NiFakeNonIviServiceTests, InitWithReturnedSessionFailsInit_AccessSession_SessionIsNotInMapAndStatusValueIsReturned)
+{
+  constexpr auto SESSION_NAME = "session";
+  constexpr auto BAD_SESSION_HANDLE = 0xDEADBEEF;  // Fake non-ivi uses 0xDEADBEEF as failed/null session (See functions.py).
+  constexpr auto FAILED_INIT = -1;
+  EXPECT_CALL(library_, InitWithReturnedSession(StrEq(SESSION_NAME)))
+      .WillOnce(Return(BAD_SESSION_HANDLE));
+  ::grpc::ServerContext context;
+  InitWithReturnedSessionRequest request;
+  InitWithReturnedSessionResponse response;
+  request.set_handle_name(SESSION_NAME);
+  service_.InitWithReturnedSession(&context, &request, &response);
+
+  const auto accessed_handle = resource_repository_->access_session(response.handle().id(), "");
+
+  EXPECT_EQ(FAILED_INIT, response.status());
+  EXPECT_EQ(0, accessed_handle);
+}
+
+TEST_F(NiFakeNonIviServiceTests, GetStringAsReturnedValue_ReturnsString)
+{
+  constexpr auto STRING_VAL = "Hello Returned World!";
+  EXPECT_CALL(library_, GetStringAsReturnedValue(_))
+      .WillOnce(Return(STRING_VAL));
+  ::grpc::ServerContext context;
+  GetStringAsReturnedValueRequest request;
+  GetStringAsReturnedValueResponse response;
+  service_.GetStringAsReturnedValue(&context, &request, &response);
+
+  EXPECT_EQ(std::string(STRING_VAL), response.string_out());
+  EXPECT_EQ(kDriverSuccess, response.status());
+}
+
+TEST_F(NiFakeNonIviServiceTests, GetStringAsReturnedValueReturnsNull_ReturnsErrorAndEmptryString)
+{
+  constexpr auto FAILED_GET = -1;
+  EXPECT_CALL(library_, GetStringAsReturnedValue(_))
+      .WillOnce(Return(nullptr));
+  ::grpc::ServerContext context;
+  GetStringAsReturnedValueRequest request;
+  GetStringAsReturnedValueResponse response;
+  service_.GetStringAsReturnedValue(&context, &request, &response);
+
+  EXPECT_THAT(response.string_out(), IsEmpty());
```

Comment:
One of these is backwards, I think normally it is expected, actual

Reply:
I do try to do `expected, actual`, since you're right it's the NI standard (at least it is in ASW). But it gets reversed when you have to do `EXPECT_THAT` because the matcher is always on the right so `actual` goes on the left. So it's:

> Expect that [actual] [is whatever]

i.e.,
> Expect that [actual] is empty.
> Expect that [actual] starts with [expected].
> Expect that [actual] container equals [expected].

It may be possible to have a hardcore rule to put [expected] in the left-hand-side in `EXPECT_THAT` statements, but  ... it would be pretty weird because you'd have to invert a lot of the asserts. And it reads weird as a sentence:

> Expect that [expected] is a superstring of [actual]

Possibly for that reason, gtest examples tend to do `actual, expected` in `EXPECT_EQ` as well. ([reference](http://google.github.io/googletest/primer.html))